---
title: "DuelingBanjos"
author: "Amy Paguirigan"
date: "04/15/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---
# Setup
## Install Required Packages
```{r}
remotes::install_github('FredHutch/tgR')
remotes::install_github('FredHutch/fh.wdlR')
```

## Load Packages
```{r}
library(tidyverse); library(tgR); library(fh.wdlR); library(aws.s3); library(httr)
```

## Set Credentials
```{r}
setCreds(path = "~/github/cred/paguirigan.R") 
```

# Pilot and Cohort 2 variant calling
PON for this entire analysis
s3://fh-pi-berger-a/tg/LILAC-FARBER/TGR-Analyses/PanelOfNormals/c3847164-8320-4c98-9e6c-4749a32fe434/LILAC-allNormals.vcf.gz
## Set up Manifest
### Pull S3 Inventory and Annotate
```{r}
bucket <- "fh-pi-berger-a"; DAG  <- "bergera"
tags <- listS3Objects(bucket = bucket)
annotations <- tgrAnnotate(DAG = DAG, harmonizedOnly = FALSE)
monsterMash <- dplyr::left_join(tags, annotations) %>% dropWhen()
```

### Filter Inventory for Workflow Input Data
In this workflow, the input data must be unmapped bams with quality control checks.  Also, to run the workflow you'll need the bed file for the dataset up in S3.  
```{r}
unique(monsterMash$workflowName)
unmappedBams <- monsterMash %>% 
  filter(workflowName %in% c("AlignedToUnMappedBam", "PairedFastqsToUnmappedBam") & 
           genomics_type == "dnaseq" & 
           workflowOutputType == "unmappedbam") %>% dropWhen(unique = T)

unmappedBams %>% group_by(workflowOutputType, dataset_qc_cohort) %>%summarize(n())
pairedSamples <- unmappedBams %>% 
  select(molecular_id, omics_sample_name, key, seq_libtype, 
         tissue_type, specimen_type, biospecimen_id, subject_id) %>% 
  filter(molecular_id != "referenceDataSet")

```

### Match Sample and Reference Datasets
```{r}
samples <- pairedSamples %>% filter(tissue_type == "lung" & specimen_type == "tumor") %>% 
  select(key, omics_sample_name, subject_id, molecular_id) %>%
  rename("sampleBamLocation" = "key", "sampleName" = "omics_sample_name")%>% filter(subject_id != "Blinded Duplicate")

references <- pairedSamples %>% filter(tissue_type == "peripheralblood") %>% 
  select(key, omics_sample_name, subject_id, molecular_id) %>% 
  rename("refBamLocation" = "key", "referenceName" = "omics_sample_name", "ref_molecular_id" = "molecular_id") %>% filter(subject_id != "Blinded Duplicate")

matched <- inner_join(samples, references, by = "subject_id") # %>% filter(subject_id != "blinded")# Check here that all is as it should be.  
matched$sampleBamLocation <- paste0("s3://fh-pi-berger-a/", matched$sampleBamLocation)
matched$refBamLocation <- paste0("s3://fh-pi-berger-a/", matched$refBamLocation)

```

### Prefilter manifest if needed
```{r}
manifest <- matched
manifest <- manifest[1,]
```

### Name and Ship Batch File
```{r}
repoName <- "FredHutch/tg-wdl-LILAC-workflow"

batchFileName <- paste0("cromwell-manifests/", repoName, "/", format(Sys.Date(), "%Y-%m-%d-"), "LILAC-75-pairs.tsv")
paste0("s3://fh-pi-paguirigan-a-genomicsrepo/", batchFileName)
s3write_using(manifest,
              FUN = write.table, quote = F, row.names = F, sep = "\t",
              object = batchFileName,
              bucket = "fh-pi-paguirigan-a-genomicsrepo")

```
## Cromwell Setup
### Server setup
```{r}
cromwellCreate(FredHutchId = "apaguiri", port = "8000",
        pathToServerLogs = " ~/cromwell/partlyCloudyConfig/server-logs/partly-v49-%A.txt",
        pathToServerScript = "~/cromwell/partlyCloudyConfig/partlyCloudy.sh",
        pathToParams = "~/cromwell/partlyCloudyConfig/cromwellParams.sh")

setCromwellURL(FredHutchId = "apaguiri", jobId = "47006887", port = "8000")

Sys.setenv("CROMWELLURL" = "http://gizmoj7:8000")
Sys.getenv("CROMWELLURL")
```


### Submit Job to Cromwell

Note: setwd to location of workflow files
Validate workflow
```{r}
valid <- womtoolValidate(WDL = "duelingBanjos_GATK4.1.4.0.wdl"); valid[["errors"]]; valid$isRunnableWorkflow
```

```{r}
setwd("~/OneDrive - Fred Hutchinson Cancer Research Center/Analysis/Berger-LILAC/ProductionWork/tg-wdl-LILAC-workflow/GATK4.1.4.0")
thisJob <- cromwellSubmitBatch(WDL = "duelingBanjos_GATK4.1.4.0.wdl",
                    Params = "duel_parameters_hg19-homosapiens.json",
                    Batch = "duel_batch_hg19-homosapiens-75-pairs.json",
                    Options = "workflow-options.json",
                    Labels = data.frame("workflowType" = "all 75 pairs"))
thisOne <- thisJob$id; thisOne
```

## Monitor Jobs
See all thejobs this server has run in the past (x) days
```{r}
afew <- 2
jobs <- cromwellJobs(days = afew) %>% arrange(desc(start))
jobs <- cromwellJobs()
```

Isolate one workflow
```{r}

thisOne <- jobs$workflow_id[1]
```

Historical job tracking
```{r}
thisManifest <- "cromwell-manifests/2020-04-17-duel-4.1.4.0-test1.tsv"
thisOne <- "91291e5c-8c64-4df4-b1e1-ae636c9f2f31"

thisOne <- "d68f8dce-cd25-41cb-8fb0-052b1095cc7a" # first submission of 59 pairs with all comers PON
thisOne <- "c8249745-6318-4f21-b463-d5e160d8dd52" # The copy outputs job

thisManifest <-  "cromwell-manifests/FredHutch/tg-wdl-LILAC-workflow/2020-05-31-LILAC-75-pairs.tsv"
thisOne <- "e4beab94-a76b-49a5-b486-437284ea8dae" # first attempt at all 75 pairs
```

Monitor Running Workflows
```{r}
w <- cromwellWorkflow(thisOne); w$status
c <- cromwellCall(thisOne); c %>% group_by(executionStatus, callName) %>% summarize(status = n())
c %>% group_by(executionStatus, callName) %>% filter(executionStatus != "Done") %>% summarize(status = n())
ca <- cromwellCache(thisOne); ca %>% group_by(callCaching.hit, callName) %>% summarize(hits = n())
ca %>% group_by(callCaching.hit) %>% summarize(hits = n())
f <- cromwellFailures(thisOne); f
```

Find out more about call caching and calls
```{r}
butWhy <- left_join(cromwellCall(thisOne) %>% mutate_all(as.character), 
                    cromwellCache(thisOne) %>% mutate_all(as.character)); butWhy %>% group_by(callName, executionStatus, callCaching.hit) %>% summarize(hits = n()) %>% arrange(desc(executionStatus))
```

Stop a workflow or find more about it
```{r}
abort <- cromwellAbort(thisOne)
WTF <- cromwellGlob(thisOne); WTF$failures
cromwellTiming(thisOne)
```

## Workflow Outputs
```{r}
out <- cromwellOutputs(thisOne) 
```



# Output Processing
## Join manifest and outputs
```{r}
out <- cromwellOutputs(thisOne)
batchFile <- manifest
batchFile <- s3read_using(
              FUN = read.table, header = T, sep = "\t",
              object = thisManifest,
              bucket = "fh-pi-paguirigan-a-genomicsrepo")
batchFile$shardIndex <- as.character(seq(from = 0, to = nrow(batchFile)-1, by = 1))

annotatedOutputs <- inner_join(batchFile, out, by = c("shardIndex"))
```

## Prep CopyNTag manifest
Column names required for copyNTag are: molecular_id, sourceFile, stage, workflowOutputType, workflowId, workflowName, destinationPrefix, destinationBucket
```{r}
copyNTag <- annotatedOutputs %>%
  select(molecular_id, workflowOutputType, pathToOutput, workflow_id, workflowName, ref_molecular_id) %>%
  rename("workflowId" = "workflow_id", "sourceFile" = "pathToOutput")
head(copyNTag)

copyNTag <- copyNTag %>% 
  mutate(destinationPrefix = 
           gsub("/fh/scratch/delete90/paguirigan_a/cromwell-executions/",
                "tg/LILAC-FARBER/TGR-Analyses/", copyNTag$sourceFile))

copyNTag$molecular_id <- as.character(copyNTag$molecular_id)
copyNTag <- copyNTag %>% 
  mutate(destinationPrefix = str_replace(
    destinationPrefix, 
    "call-[^/]*/shard-.*/execution", 
    molecular_id)) # Now works for cacheCopy files but be careful here. 
head(copyNTag$destinationPrefix)
tail(copyNTag$destinationPrefix)

copyNTag$destinationBucket <- "fh-pi-berger-a"
copyNTag$stage <- "processed"
copyNTag[10,]


referenceOutputs <-  c("analysisReadyRefIndex" , "analysisReadyRefBam", "refPicardQCpertarget" , "refPicardQC" ,"refBedQC")
refOutputData <- copyNTag %>% filter(workflowOutputType %in% referenceOutputs) %>% 
  select(-molecular_id) %>% rename("molecular_id" = "ref_molecular_id")
head(refOutputData)
tail(refOutputData)

sampleOutputData <- copyNTag %>% filter(!workflowOutputType %in% referenceOutputs) %>% select(-ref_molecular_id)
merged <- rbind(sampleOutputData, refOutputData)
```

## Create Manifest
```{r}
copyManifest <- merged %>% 
  select(sourceFile, molecular_id, stage, workflowId, workflowName, workflowOutputType, destinationPrefix, destinationBucket)
```


## Ship Batch File
```{r}
repoName <- "FredHutch/tg-wdl-copyNTag"
batchFileName <- paste0("cromwell-manifests/",repoName, "/", format(Sys.Date(), "%Y-%m-%d-"), "75-LILAC-pairs-fix.tsv")
paste0("s3://fh-pi-paguirigan-a-genomicsrepo/", batchFileName)
s3write_using(copyManifest,
              FUN = write.table, quote = F, row.names = F, sep = "\t",
              object = batchFileName,
              bucket = "fh-pi-paguirigan-a-genomicsrepo")
thisManifest <- "s3://fh-pi-paguirigan-a-genomicsrepo/cromwell-manifests/FredHutch/tg-wdl-copyNTag/2020-05-31-75-LILAC-pairs-fix.tsv"
```

## Validate and Submit Job
```{r}
Sys.setenv("CROMWELLURL" = "http://gizmog3:8000")

setwd("~/OneDrive - Fred Hutchinson Cancer Research Center/Analysis/Berger-LILAC/ProductionWork/tg-wdl-LILAC-workflow/GATK4.1.4.0/post-copyTag")
womtoolValidate("copyNtag.wdl")

thisJob <- cromwellSubmitBatch(
  WDL = "copyNtag.wdl",
  Params = "copyNtag-inputs.json",
  Options = "workflow-options.json",
  Labels = data.frame("workflowType" = "LILAC"))

thisOne <- thisJob$id; thisOne
```

```{r}
thisManifest <- "s3://fh-pi-paguirigan-a-genomicsrepo/cromwell-manifests/FredHutch/tg-wdl-copyNTag/2020-05-31-75-LILAC-pairs.tsv"
thisOne <- "87cffbd3-fe9d-4bc7-afd5-aed842a1ed50"
thisOne <- "aea721c9-8989-449f-8885-50f79b176723"
thisOne <- "f25fc2a9-4afe-417a-8c33-00c8818d7af4" # Fix tags
thisOne <- "f1d1ec2f-730b-439f-b2b7-1ac928470c89" # path fixed.  sheesh
thisOne <- "663cbe6b-4d06-430b-972f-924dbc753c07" #OMFG
thisOne <- "090e60d1-ff78-4176-bbe4-da8247ba23c9" # resubmit to fnish
thisOne <- "ea536d79-6c81-4493-b416-7d033abdb0eb"
```


## Workflow level metadata
```{r}
w <- cromwellWorkflow(thisOne); w$status
WTF <- cromwellGlob(thisOne); WTF[["failures"]]
c <- cromwellCall(thisOne); c %>% group_by(executionStatus, callName) %>% summarize(status = n())
ca <- cromwellCache(thisOne); ca %>% group_by(callCaching.hit, callName) %>% summarize(hits = n())
f <- cromwellFailures(thisOne); f
cromwellAbort(thisOne)
```


## Confirm
```{r}
Bucket <- "fh-pi-berger-a"
Dag <- "bergera"
tags <- listS3Objects(bucket = Bucket)
annotations <- tgrAnnotate(DAG = Dag)
monsterMash <- left_join(annotations, tags)

unique(monsterMash$workflowName)
monsterMash %>% filter(workflowName == "HybridCap_BWA_Mutect2_Strelka_AnnotatedVariants") %>% distinct(workflowID)
monsterMash %>% filter(workflowName == "HybridCap_BWA_Mutect2_Strelka_AnnotatedVariants" & 
                         workflowID == "e4beab94-a76b-49a5-b486-437284ea8dae") %>% distinct(workflowOutputType)

```


